zookeeper-server/src/main/java/org/apache/zookeeper/backup/NFSBackupStorage.java

```diff
@@ -0,0 +1,324 @@
+    } catch (FileNotFoundException e) {
+      if (!backupFile.exists() || backupFile.isDirectory()) {
+        throw new BackupException("The backed up file with the file path " + backupFilePath
+            + " does exists or is a directory, could not copy to local storage.");
+      } else {
+        throw new IOException("Cannot open source file in backup storage" + backupFilePath
+            + " , cannot be copied to local storage.");
+      }
+    }
+
+    // Prepare in local storage: create parent dirs if not exist, create the destination file
+    destFile.getParentFile().mkdirs();
+    if (!destFile.createNewFile()) {
+      if (destFile.exists()) {
+        throw new BackupException("A local storage file with the file path " + destFile.getPath()
+            + " already exists, could not create a new one.");
+      } else {
+        throw new IOException("Failed to create a file at destination path: " + destFile.getPath()
+            + ", stopping the streaming.");
+      }
+    }
+
+    // Create output stream on the destination file and start streaming data
+    try {
+      outputStream = new FileOutputStream(destFile);
+
+      byte[] buffer = new byte[1024];
+      int lengthRead;
+      while ((lengthRead = inputStream.read(buffer)) > 0) {
+        outputStream.write(buffer, 0, lengthRead);
+        outputStream.flush();
+      }
+    } catch (FileNotFoundException e) {
+      throw new IOException(
+          "The destination local file " + destFile.getPath() + " could not be opened for write.");
+    } catch (IOException e) {
+      throw new IOException(
+          "IOException is caught during data streaming while copying file " + backupFilePath
+              + " to local storage. The message of the IOException is: " + e.getMessage());
+    } finally {
+      inputStream.close();
+      if (outputStream != null) {
+        outputStream.close();
+      }
+    }
+  }
+
+  @Override
+  public void delete(File fileToDelete) throws IOException {
+    String backupFilePath = constructBackupFilePath(fileToDelete.getPath());
+    try {
+      if (!Files.deleteIfExists(Paths.get(backupFilePath))) {
+        throw new BackupException(
+            "File " + fileToDelete.getName() + "does not exist, cannot delete this file.");
+      }
+    } catch (DirectoryNotEmptyException e) {
+      throw new BackupException(
+          "Directory " + fileToDelete.getName() + "is not empty, cannot be deleted.");
+    }
+  }
+
+  /**
+   * Construct the file path in backup storage using file name
+   * @param fileName The file name for the file
+   * @return The file path to the location where the file will be stored in backup storage
+   */
+  private String constructBackupFilePath(String fileName) {
+    String backupFileName = fileName;
+
+    //Remove the temporary file name prefix in order to determine file type
+    if (fileName.startsWith(TMP_FILE_PREFIX)) {
+      backupFileName = fileName.replace(TMP_FILE_PREFIX, "");
```

Comment:
Can you use substring() instead?
https://stackoverflow.com/questions/30549573/is-substring-or-replace-faster-to-remove-the-last-character-in-a-string